# 2. Times Higher Education World University Rankings Exploratory Analysis

## Importing libraries

In [60]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

## Importing dataset

In [61]:
# Set path variable

path = r'C:\Users\HP\Documents\CareerFoundry\Data Immersion\Achievement 6\World University Rankings Analysis'

In [62]:
# Import dataset

df_times = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'timesData.csv'))

In [63]:
# Check to see if it imported correctly

df_times.head(10)

world_rank                        university_name                   country  \
0          1                     Harvard University  United States of America   
1          2     California Institute of Technology  United States of America   
2          3  Massachusetts Institute of Technology  United States of America   
3          4                    Stanford University  United States of America   
4          5                   Princeton University  United States of America   
5          6                University of Cambridge            United Kingdom   
6          6                   University of Oxford            United Kingdom   
7          8     University of California, Berkeley  United States of America   
8          9                Imperial College London            United Kingdom   
9         10                        Yale University  United States of America   

   teaching international  research  citations income total_score  \
0      99.7          72.4      98.7       98.8   34.5        96.1   
1      97.7          54.6      98.0       99.9   83.7        96.0   
2      97.8          82.3      91.4       99.9   87.5        95.6   
3      98.3          29.5      98.1       99.2   64.3        94.3   
4      90.9          70.3      95.4       99.9      -        94.2   
5      90.5          77.7      94.1       94.0   57.0        91.2   
6      88.2          77.2      93.9       95.1   73.5        91.2   
7      84.2          39.6      99.3       97.8      -        91.1   
8      89.2          90.0      94.5       88.3   92.9        90.6   
9      92.1          59.2      89.7       91.5      -        89.5   

  num_students  student_staff_ratio international_students female_male_ratio  \
0       20,152                  8.9                    25%               NaN   
1        2,243                  6.9                    27%           33 : 67   
2       11,074                  9.0                    33%           37 : 63   
3       15,596                  7.8                    22%           42 : 58   
4        7,929                  8.4                    27%           45 : 55   
5       18,812                 11.8                    34%           46 : 54   
6       19,919                 11.6                    34%           46 : 54   
7       36,186                 16.4                    15%           50 : 50   
8       15,060                 11.7                    51%           37 : 63   
9       11,751                  4.4                    20%           50 : 50   

   year  
0  2011  
1  2011  
2  2011  
3  2011  
4  2011  
5  2011  
6  2011  
7  2011  
8  2011  
9  2011

In [64]:
# Check dimensions of dataframe

df_times.shape

(2603, 14)

## Data Cleaning

### Dropping Irrelevant Columns

In [65]:
# Check the column names

df_times.columns

Index(['world_rank', 'university_name', 'country', 'teaching', 'international',
       'research', 'citations', 'income', 'total_score', 'num_students',
       'student_staff_ratio', 'international_students', 'female_male_ratio',
       'year'],
      dtype='object')

It looks like most columns would be relevant for analysis. The rankings look to be based on certain criteria: teaching, international, research, and citations. Extra additional information such as number of students, student staff ratio, and female to male ratio may lead to more insights. 

### Renaming Columns

In [66]:
# Rename certain columns to be more intuitive

df_times.rename(columns = {'world_rank' : 'world_rank_Times',
                           'teaching' : 'teaching_score',
                           'international' : 'international_outlook_score',
                           'research' : 'research_score',
                           'citations' : 'citations_score',
                           'income' : 'income_score',
                           'total_score' : 'overall_score_Times'}, inplace = True)

In [67]:
# Check column names to see if they were replaced correctly

df_times.columns

Index(['world_rank_Times', 'university_name', 'country', 'teaching_score',
       'international_outlook_score', 'research_score', 'citations_score',
       'income_score', 'overall_score_Times', 'num_students',
       'student_staff_ratio', 'international_students', 'female_male_ratio',
       'year'],
      dtype='object')

### Changing a Variable Data Type

In [68]:
df_times.dtypes

world_rank_Times                object
university_name                 object
country                         object
teaching_score                 float64
international_outlook_score     object
research_score                 float64
citations_score                float64
income_score                    object
overall_score_Times             object
num_students                    object
student_staff_ratio            float64
international_students          object
female_male_ratio               object
year                             int64
dtype: object

Certain columns have strange data types. This may be because of missing values. 

#### Working with the 'international_outlook_score' column

In [69]:
# Check for missing or strange values in the 'international' column

df_times['international_outlook_score'].value_counts(dropna = False)

29.6    10
20.7    10
46.8     9
34.3     9
-        9
        ..
40.7     1
68.8     1
22.8     1
38.9     1
14.9     1
Name: international_outlook_score, Length: 804, dtype: int64

There are 9 rows in the international_outlook_score column that have blanks ('-'). These can be replaced with NaN values. 

In [70]:
# Replace the '-' with NaN values.

df_times['international_outlook_score'] = df_times['international_outlook_score'].replace('-', np.nan)

In [71]:
# Check the number of NaN values using isnull() and sum()

df_times['international_outlook_score'].isnull().sum()

9

#### Working with the 'income_score' column

In [72]:
# Check for missing or strange values in the 'income' column

df_times['income_score'].value_counts(dropna = False)

-        218
100.0     68
28.0      26
31.1      20
28.8      19
        ... 
55.8       1
89.0       1
73.5       1
84.8       1
89.7       1
Name: income_score, Length: 613, dtype: int64

There are 218 rows in the income_score column that have blanks ('-'). These can be replaced with NaN values. 

In [73]:
# Replace the '-' with NaN values.

df_times['income_score'] = df_times['income_score'].replace('-', np.nan)

In [74]:
# Check the number of NaN values using isnull() and sum()

df_times['income_score'].isnull().sum()

218

#### Working with the 'overall_score_Times' column

In [75]:
# Check for missing or strange values in the 'overall_score' column

df_times['overall_score_Times'].value_counts(dropna = False)

-       1402
49.0      13
51.1      12
46.6      11
51.2      10
        ... 
63.8       1
87.3       1
90.2       1
90.7       1
96.1       1
Name: overall_score_Times, Length: 415, dtype: int64

There are 1402 rows in the overall_score_Times column that have blanks ('-'). These can be replaced with NaN values. 

In [76]:
# Replace the '-' with NaN values.

df_times['overall_score_Times'] = df_times['overall_score_Times'].replace('-', np.nan)

In [77]:
# Check the number of NaN values using isnull() and sum()

df_times['overall_score_Times'].isnull().sum()

1402

#### Working with the 'num_students' column

In [78]:
# Check for missing or strange values in the 'num_students' column

df_times['num_students'].value_counts(dropna = False)

NaN       59
20,152     6
23,280     6
23,144     6
17,581     6
          ..
22,422     1
6,898      1
10,546     1
10,697     1
10,117     1
Name: num_students, Length: 795, dtype: int64

In [79]:
# Check the number of NaN values using isnull() and sum()

df_times['num_students'].isnull().sum()

59

There are 59 rows in the num_students column with NaN values. 

#### Working with the 'student_staff_ratio' column

In [80]:
# Check for missing or strange values in the 'student_staff ratio' column

df_times['student_staff_ratio'].value_counts(dropna = False)

NaN     59
15.9    38
17.4    32
13.0    32
25.9    28
        ..
4.0      1
20.8     1
49.7     1
29.6     1
28.7     1
Name: student_staff_ratio, Length: 309, dtype: int64

In [81]:
# Check the number of NaN values using isnull() and sum()

df_times['student_staff_ratio'].isnull().sum()

59

There are 59 rows in the student_staff_ratio column with NaN values. 

#### Working with the 'international_students' column

In [82]:
# Check for missing or strange values in the 'international_students' column

df_times['international_students'].value_counts(dropna = False)

7%     142
10%    133
9%     130
5%     120
8%     119
12%    104
15%    103
11%    100
16%     90
1%      80
14%     80
17%     79
18%     78
4%      77
13%     76
6%      74
NaN     67
25%     67
20%     66
19%     66
3%      59
22%     56
2%      54
21%     51
23%     50
28%     48
27%     45
26%     45
33%     34
35%     30
0%      26
34%     26
38%     25
24%     24
37%     23
30%     22
29%     18
36%     18
39%     17
31%     13
32%     12
43%      9
44%      6
46%      6
40%      6
47%      6
51%      6
54%      6
48%      5
63%      2
52%      1
50%      1
82%      1
42%      1
Name: international_students, dtype: int64

In [83]:
# Check the number of NaN values using isnull() and sum()

df_times['international_students'].isnull().sum()

67

There are 67 rows in the international_students column with NaN values. 

#### Working with the 'female_male_ratio' column

In [84]:
# Check for missing or strange values in the 'female_male_ratio' column

df_times['female_male_ratio'].value_counts(dropna = False)

NaN        233
54 : 46    185
52 : 48    151
53 : 47    138
55 : 45    135
          ... 
11 : 89      1
68 : 32      1
1 : 99       1
74 : 26      1
9 : 91       1
Name: female_male_ratio, Length: 70, dtype: int64

In [85]:
# Check the number of NaN values using isnull() and sum()

df_times['female_male_ratio'].isnull().sum()

233

There are 233 rows in the student_staff_ratio column with NaN values. However, there are also 3 rows with '-' that should be replaced with NaNs to be consistent.

In [86]:
# Replace the '-' with NaN values.

df_times['female_male_ratio'] = df_times['female_male_ratio'].replace('-', np.nan)

In [87]:
# Check the number of NaN values using isnull() and sum()

df_times['female_male_ratio'].isnull().sum()

236

There are now 236 NaN values, which indicates that the 3 '-' values were changes correctly. 

### Addressing Mixed Data Columns

In [88]:
# Check for mixed data types

for col in df_times.columns.tolist():
  weird = (df_times[[col]].applymap(type) != df_times[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_times[weird]) > 0:
    print (col)

international_outlook_score
income_score
overall_score_Times
num_students
international_students
female_male_ratio


Even after replacing blanks, there are still columns that have mixed data types. After replacing the blanks with NaN's using np.Nan, these should be recognized as floats. As such, the other columns should also be changed to floats so that statistics can be calculated for them.

In [89]:
df_times.dtypes

world_rank_Times                object
university_name                 object
country                         object
teaching_score                 float64
international_outlook_score     object
research_score                 float64
citations_score                float64
income_score                    object
overall_score_Times             object
num_students                    object
student_staff_ratio            float64
international_students          object
female_male_ratio               object
year                             int64
dtype: object

In [90]:
# Change the international, income, overall_score, and num_students columns

df_times['international_outlook_score'] = df_times['international_outlook_score'].astype('float64')
df_times['income_score'] = df_times['income_score'].astype('float64')
df_times['overall_score_Times'] = df_times['overall_score_Times'].astype('float64')

In [91]:
df_times.dtypes

world_rank_Times                object
university_name                 object
country                         object
teaching_score                 float64
international_outlook_score    float64
research_score                 float64
citations_score                float64
income_score                   float64
overall_score_Times            float64
num_students                    object
student_staff_ratio            float64
international_students          object
female_male_ratio               object
year                             int64
dtype: object

#### Working with the 'num_students' column

In [92]:
# The num_students column is viewed as strings because there are both blanks and some numbers (integers).
# The best practice would be to change the entire column to a float since blanks can be turned into NaN's and the integers can just be turned into floats
# First, convert the entire column to strings

df_times['num_students'] = df_times['num_students'].astype('str')

In [93]:
# Check the datatype

df_times['num_students'].dtype

dtype('O')

In [94]:
# Create an empty column that will be filled based on values in the num_students column criteria

results_num_stud = []

for value in df_times['num_students']:
    if value == 'nan':
        results_num_stud.append(np.nan)
    else:
        results_num_stud.append(value.replace(',', ''))

df_times['num_stud_float'] = results_num_stud

df_times.head(20)

world_rank_Times                                    university_name  \
0                 1                                 Harvard University   
1                 2                 California Institute of Technology   
2                 3              Massachusetts Institute of Technology   
3                 4                                Stanford University   
4                 5                               Princeton University   
5                 6                            University of Cambridge   
6                 6                               University of Oxford   
7                 8                 University of California, Berkeley   
8                 9                            Imperial College London   
9                10                                    Yale University   
10               11              University of California, Los Angeles   
11               12                              University of Chicago   
12               13                           Johns Hopkins University   
13               14                                 Cornell University   
14               15  ETH Zurich – Swiss Federal Institute of Techno...   
15               15                             University of Michigan   
16               17                              University of Toronto   
17               18                                Columbia University   
18               19                         University of Pennsylvania   
19               20                         Carnegie Mellon University   

                     country  teaching_score  international_outlook_score  \
0   United States of America            99.7                         72.4   
1   United States of America            97.7                         54.6   
2   United States of America            97.8                         82.3   
3   United States of America            98.3                         29.5   
4   United States of America            90.9                         70.3   
5             United Kingdom            90.5                         77.7   
6             United Kingdom            88.2                         77.2   
7   United States of America            84.2                         39.6   
8             United Kingdom            89.2                         90.0   
9   United States of America            92.1                         59.2   
10  United States of America            83.0                         48.1   
11  United States of America            79.1                         62.8   
12  United States of America            80.9                         58.5   
13  United States of America            82.2                         62.4   
14               Switzerland            77.5                         93.7   
15  United States of America            83.9                         53.3   
16                    Canada            75.8                          NaN   
17  United States of America            73.8                         90.9   
18  United States of America            71.8                         32.9   
19  United States of America            70.3                         39.1   

    research_score  citations_score  income_score  overall_score_Times  \
0             98.7             98.8          34.5                 96.1   
1             98.0             99.9          83.7                 96.0   
2             91.4             99.9          87.5                 95.6   
3             98.1             99.2          64.3                 94.3   
4             95.4             99.9           NaN                 94.2   
5             94.1             94.0          57.0                 91.2   
6             93.9             95.1          73.5                 91.2   
7             99.3             97.8           NaN                 91.1   
8             94.5             88.3          92.9                 90.6   
9             89.7             91.5           NaN                 89.5   
10            92

In [95]:
# Replace the original column with the cleaned up values

df_times['num_students'] = df_times['num_stud_float']

df_times

world_rank_Times                        university_name  \
0                   1                     Harvard University   
1                   2     California Institute of Technology   
2                   3  Massachusetts Institute of Technology   
3                   4                    Stanford University   
4                   5                   Princeton University   
...               ...                                    ...   
2598          601-800                    Yeungnam University   
2599          601-800            Yıldız Technical University   
2600          601-800               Yokohama City University   
2601          601-800           Yokohama National University   
2602          601-800                     Yuan Ze University   

                       country  teaching_score  international_outlook_score  \
0     United States of America            99.7                         72.4   
1     United States of America            97.7                         54.6   
2     United States of America            97.8                         82.3   
3     United States of America            98.3                         29.5   
4     United States of America            90.9                         70.3   
...                        ...             ...                          ...   
2598               South Korea            18.6                         24.3   
2599                    Turkey            14.5                         14.9   
2600                     Japan            24.0                         16.1   
2601                     Japan            20.1                         23.3   
2602                    Taiwan            16.2                         17.7   

      research_score  citations_score  income_score  overall_score_Times  \
0               98.7             98.8          34.5                 96.1   
1               98.0             99.9          83.7                 96.0   
2               91.4             99.9          87.5                 95.6   
3               98.1             99.2          64.3                 94.3   
4               95.4             99.9           NaN                 94.2   
...              ...              ...           ...                  ...   
2598            10.9             26.5          35.4                  NaN   
2599             7.6             19.3          44.0                  NaN   
2600            10.2             36.4          37.9                  NaN   
2601            16.0             13.5          40.4                  NaN   
2602            18.3             28.6          39.8                  NaN   

     num_students  student_staff_ratio international_students  \
0           20152                  8.9                    25%   
1            2243                  6.9                    27%   
2           11074                  9.0                    33%   
3           15596                  7.8                    22%   
4            7929                  8.4                    27%   
...           ...                  ...                    ...   
2598        21958                 15.3                     3%   
2599        31268                 28.7                     2%   
2600         4122                  3.7                     3%   
2601        10117                 12.1                     8%   
2602         8663                 20.6                     4%   

     female_male_ratio  year num_stud_float  
0                  NaN  2011          20152  
1              33 : 67  2011           2243  
2              37 : 63  2011          11074  
3              42 : 58  2011          15596  
4              45 : 55  2011           7929  
...                ...   ...            ...  
2598           48 : 52  2016          21958  
2599           36 : 64  2016          31268  
2600               NaN  2016           4122  
2601           28 : 72  2016          10117  
2602           43 : 57  2016           8663  

[2603 rows x 15 columns]

In [96]:
# Drop the newly created column

df_times = df_times.drop(columns = ['num_stud_float'])

df_times

world_rank_Times                        university_name  \
0                   1                     Harvard University   
1                   2     California Institute of Technology   
2                   3  Massachusetts Institute of Technology   
3                   4                    Stanford University   
4                   5                   Princeton University   
...               ...                                    ...   
2598          601-800                    Yeungnam University   
2599          601-800            Yıldız Technical University   
2600          601-800               Yokohama City University   
2601          601-800           Yokohama National University   
2602          601-800                     Yuan Ze University   

                       country  teaching_score  international_outlook_score  \
0     United States of America            99.7                         72.4   
1     United States of America            97.7                         54.6   
2     United States of America            97.8                         82.3   
3     United States of America            98.3                         29.5   
4     United States of America            90.9                         70.3   
...                        ...             ...                          ...   
2598               South Korea            18.6                         24.3   
2599                    Turkey            14.5                         14.9   
2600                     Japan            24.0                         16.1   
2601                     Japan            20.1                         23.3   
2602                    Taiwan            16.2                         17.7   

      research_score  citations_score  income_score  overall_score_Times  \
0               98.7             98.8          34.5                 96.1   
1               98.0             99.9          83.7                 96.0   
2               91.4             99.9          87.5                 95.6   
3               98.1             99.2          64.3                 94.3   
4               95.4             99.9           NaN                 94.2   
...              ...              ...           ...                  ...   
2598            10.9             26.5          35.4                  NaN   
2599             7.6             19.3          44.0                  NaN   
2600            10.2             36.4          37.9                  NaN   
2601            16.0             13.5          40.4                  NaN   
2602            18.3             28.6          39.8                  NaN   

     num_students  student_staff_ratio international_students  \
0           20152                  8.9                    25%   
1            2243                  6.9                    27%   
2           11074                  9.0                    33%   
3           15596                  7.8                    22%   
4            7929                  8.4                    27%   
...           ...                  ...                    ...   
2598        21958                 15.3                     3%   
2599        31268                 28.7                     2%   
2600         4122                  3.7                     3%   
2601        10117                 12.1                     8%   
2602         8663                 20.6                     4%   

     female_male_ratio  year  
0                  NaN  2011  
1              33 : 67  2011  
2              37 : 63  2011  
3              42 : 58  2011  
4              45 : 55  2011  
...                ...   ...  
2598           48 : 52  2016  
2599           36 : 64  2016  
2600               NaN  2016  
2601           28 : 72  2016  
2602           43 : 57  2016  

[2603 rows x 14 columns]

In [97]:
#Check the column to see if it is a float type column

df_times['num_students'].dtype

dtype('O')

In [98]:
#Convert the column to a float type column

df_times['num_students'] = df_times['num_students'].astype('float')

df_times['num_students'].dtype

dtype('float64')

#### Working with the 'international_students' column to change values from percentages to decimals

In [99]:
# For the international_students column, the majority of the values are percentages, which Python views as strings (hence the column having an object datatype)
# Need to convert these percentages to decimals to be consistent and make the column a float64 column in the end
# First, convert all the values in the column to strings

df_times['international_students'] = df_times['international_students'].astype('str')

In [100]:
df_times['international_students'].dtype

dtype('O')

In [101]:
# Create an empty column that will be filled in based on values in the international_students column criteria

results_int_stud = []

for value in df_times['international_students']:
    if value == 'nan' or value == '':
        results_int_stud.append(np.nan)
    else:
        results_int_stud.append(float(value.replace('%', '')) / 100)
        
df_times['int_stud_dec'] = results_int_stud
        
df_times

world_rank_Times                        university_name  \
0                   1                     Harvard University   
1                   2     California Institute of Technology   
2                   3  Massachusetts Institute of Technology   
3                   4                    Stanford University   
4                   5                   Princeton University   
...               ...                                    ...   
2598          601-800                    Yeungnam University   
2599          601-800            Yıldız Technical University   
2600          601-800               Yokohama City University   
2601          601-800           Yokohama National University   
2602          601-800                     Yuan Ze University   

                       country  teaching_score  international_outlook_score  \
0     United States of America            99.7                         72.4   
1     United States of America            97.7                         54.6   
2     United States of America            97.8                         82.3   
3     United States of America            98.3                         29.5   
4     United States of America            90.9                         70.3   
...                        ...             ...                          ...   
2598               South Korea            18.6                         24.3   
2599                    Turkey            14.5                         14.9   
2600                     Japan            24.0                         16.1   
2601                     Japan            20.1                         23.3   
2602                    Taiwan            16.2                         17.7   

      research_score  citations_score  income_score  overall_score_Times  \
0               98.7             98.8          34.5                 96.1   
1               98.0             99.9          83.7                 96.0   
2               91.4             99.9          87.5                 95.6   
3               98.1             99.2          64.3                 94.3   
4               95.4             99.9           NaN                 94.2   
...              ...              ...           ...                  ...   
2598            10.9             26.5          35.4                  NaN   
2599             7.6             19.3          44.0                  NaN   
2600            10.2             36.4          37.9                  NaN   
2601            16.0             13.5          40.4                  NaN   
2602            18.3             28.6          39.8                  NaN   

      num_students  student_staff_ratio international_students  \
0          20152.0                  8.9                    25%   
1           2243.0                  6.9                    27%   
2          11074.0                  9.0                    33%   
3          15596.0                  7.8                    22%   
4           7929.0                  8.4                    27%   
...            ...                  ...                    ...   
2598       21958.0                 15.3                     3%   
2599       31268.0                 28.7                     2%   
2600        4122.0                  3.7                     3%   
2601       10117.0                 12.1                     8%   
2602        8663.0                 20.6                     4%   

     female_male_ratio  year  int_stud_dec  
0                  NaN  2011          0.25  
1              33 : 67  2011          0.27  
2              37 : 63  2011          0.33  
3              42 : 58  2011          0.22  
4              45 : 55  2011          0.27  
...                ...   ...           ...  
2598           48 : 52  2016          0.03  
2599           36 : 64  2016          0.02  
2600               NaN  2016          0.03  
2601           28 : 72  2016          0.08  
2602           43 : 57  2016          0.04  

[2603 rows x 15 columns]

In [102]:
# Replace the original column with the cleaned up values

df_times['international_students'] = df_times['int_stud_dec']

df_times

world_rank_Times                        university_name  \
0                   1                     Harvard University   
1                   2     California Institute of Technology   
2                   3  Massachusetts Institute of Technology   
3                   4                    Stanford University   
4                   5                   Princeton University   
...               ...                                    ...   
2598          601-800                    Yeungnam University   
2599          601-800            Yıldız Technical University   
2600          601-800               Yokohama City University   
2601          601-800           Yokohama National University   
2602          601-800                     Yuan Ze University   

                       country  teaching_score  international_outlook_score  \
0     United States of America            99.7                         72.4   
1     United States of America            97.7                         54.6   
2     United States of America            97.8                         82.3   
3     United States of America            98.3                         29.5   
4     United States of America            90.9                         70.3   
...                        ...             ...                          ...   
2598               South Korea            18.6                         24.3   
2599                    Turkey            14.5                         14.9   
2600                     Japan            24.0                         16.1   
2601                     Japan            20.1                         23.3   
2602                    Taiwan            16.2                         17.7   

      research_score  citations_score  income_score  overall_score_Times  \
0               98.7             98.8          34.5                 96.1   
1               98.0             99.9          83.7                 96.0   
2               91.4             99.9          87.5                 95.6   
3               98.1             99.2          64.3                 94.3   
4               95.4             99.9           NaN                 94.2   
...              ...              ...           ...                  ...   
2598            10.9             26.5          35.4                  NaN   
2599             7.6             19.3          44.0                  NaN   
2600            10.2             36.4          37.9                  NaN   
2601            16.0             13.5          40.4                  NaN   
2602            18.3             28.6          39.8                  NaN   

      num_students  student_staff_ratio  international_students  \
0          20152.0                  8.9                    0.25   
1           2243.0                  6.9                    0.27   
2          11074.0                  9.0                    0.33   
3          15596.0                  7.8                    0.22   
4           7929.0                  8.4                    0.27   
...            ...                  ...                     ...   
2598       21958.0                 15.3                    0.03   
2599       31268.0                 28.7                    0.02   
2600        4122.0                  3.7                    0.03   
2601       10117.0                 12.1                    0.08   
2602        8663.0                 20.6                    0.04   

     female_male_ratio  year  int_stud_dec  
0                  NaN  2011          0.25  
1              33 : 67  2011          0.27  
2              37 : 63  2011          0.33  
3              42 : 58  2011          0.22  
4              45 : 55  2011          0.27  
...                ...   ...           ...  
2598           48 : 52  2016          0.03  
2599           36 : 64  2016          0.02  
2600               NaN  2016          0.03  
2601           28 : 72  2016          0.08  
2602           43 : 57  2016          0.04  

[2603 rows x 15 columns]

In [103]:
# Drop the newly created column

df_times = df_times.drop(columns = ['int_stud_dec'])

df_times

world_rank_Times                        university_name  \
0                   1                     Harvard University   
1                   2     California Institute of Technology   
2                   3  Massachusetts Institute of Technology   
3                   4                    Stanford University   
4                   5                   Princeton University   
...               ...                                    ...   
2598          601-800                    Yeungnam University   
2599          601-800            Yıldız Technical University   
2600          601-800               Yokohama City University   
2601          601-800           Yokohama National University   
2602          601-800                     Yuan Ze University   

                       country  teaching_score  international_outlook_score  \
0     United States of America            99.7                         72.4   
1     United States of America            97.7                         54.6   
2     United States of America            97.8                         82.3   
3     United States of America            98.3                         29.5   
4     United States of America            90.9                         70.3   
...                        ...             ...                          ...   
2598               South Korea            18.6                         24.3   
2599                    Turkey            14.5                         14.9   
2600                     Japan            24.0                         16.1   
2601                     Japan            20.1                         23.3   
2602                    Taiwan            16.2                         17.7   

      research_score  citations_score  income_score  overall_score_Times  \
0               98.7             98.8          34.5                 96.1   
1               98.0             99.9          83.7                 96.0   
2               91.4             99.9          87.5                 95.6   
3               98.1             99.2          64.3                 94.3   
4               95.4             99.9           NaN                 94.2   
...              ...              ...           ...                  ...   
2598            10.9             26.5          35.4                  NaN   
2599             7.6             19.3          44.0                  NaN   
2600            10.2             36.4          37.9                  NaN   
2601            16.0             13.5          40.4                  NaN   
2602            18.3             28.6          39.8                  NaN   

      num_students  student_staff_ratio  international_students  \
0          20152.0                  8.9                    0.25   
1           2243.0                  6.9                    0.27   
2          11074.0                  9.0                    0.33   
3          15596.0                  7.8                    0.22   
4           7929.0                  8.4                    0.27   
...            ...                  ...                     ...   
2598       21958.0                 15.3                    0.03   
2599       31268.0                 28.7                    0.02   
2600        4122.0                  3.7                    0.03   
2601       10117.0                 12.1                    0.08   
2602        8663.0                 20.6                    0.04   

     female_male_ratio  year  
0                  NaN  2011  
1              33 : 67  2011  
2              37 : 63  2011  
3              42 : 58  2011  
4              45 : 55  2011  
...                ...   ...  
2598           48 : 52  2016  
2599           36 : 64  2016  
2600               NaN  2016  
2601           28 : 72  2016  
2602           43 : 57  2016  

[2603 rows x 14 columns]

In [104]:
#Check the column to see if it is a float type column

df_times['international_students'].dtype

dtype('float64')

In [105]:
df_times.dtypes

world_rank_Times                object
university_name                 object
country                         object
teaching_score                 float64
international_outlook_score    float64
research_score                 float64
citations_score                float64
income_score                   float64
overall_score_Times            float64
num_students                   float64
student_staff_ratio            float64
international_students         float64
female_male_ratio               object
year                             int64
dtype: object

#### Working with female_male_ratio column to change everything to decimals. 

In [106]:
# Upon insepction, some of the values in the 'female_male_ratio' column are strings (the ones with ':') but some are numbers and floats (decimals and nans).
# Change all of them to strings before applying further changes. 

df_times['female_male_ratio'] = df_times['female_male_ratio'].astype('str')

In [107]:
df_times['female_male_ratio'].dtype

dtype('O')

In [108]:
# Remove all whitespace

df_times['female_male_ratio'] = df_times['female_male_ratio'].str.replace(" ", "")

df_times

world_rank_Times                        university_name  \
0                   1                     Harvard University   
1                   2     California Institute of Technology   
2                   3  Massachusetts Institute of Technology   
3                   4                    Stanford University   
4                   5                   Princeton University   
...               ...                                    ...   
2598          601-800                    Yeungnam University   
2599          601-800            Yıldız Technical University   
2600          601-800               Yokohama City University   
2601          601-800           Yokohama National University   
2602          601-800                     Yuan Ze University   

                       country  teaching_score  international_outlook_score  \
0     United States of America            99.7                         72.4   
1     United States of America            97.7                         54.6   
2     United States of America            97.8                         82.3   
3     United States of America            98.3                         29.5   
4     United States of America            90.9                         70.3   
...                        ...             ...                          ...   
2598               South Korea            18.6                         24.3   
2599                    Turkey            14.5                         14.9   
2600                     Japan            24.0                         16.1   
2601                     Japan            20.1                         23.3   
2602                    Taiwan            16.2                         17.7   

      research_score  citations_score  income_score  overall_score_Times  \
0               98.7             98.8          34.5                 96.1   
1               98.0             99.9          83.7                 96.0   
2               91.4             99.9          87.5                 95.6   
3               98.1             99.2          64.3                 94.3   
4               95.4             99.9           NaN                 94.2   
...              ...              ...           ...                  ...   
2598            10.9             26.5          35.4                  NaN   
2599             7.6             19.3          44.0                  NaN   
2600            10.2             36.4          37.9                  NaN   
2601            16.0             13.5          40.4                  NaN   
2602            18.3             28.6          39.8                  NaN   

      num_students  student_staff_ratio  international_students  \
0          20152.0                  8.9                    0.25   
1           2243.0                  6.9                    0.27   
2          11074.0                  9.0                    0.33   
3          15596.0                  7.8                    0.22   
4           7929.0                  8.4                    0.27   
...            ...                  ...                     ...   
2598       21958.0                 15.3                    0.03   
2599       31268.0                 28.7                    0.02   
2600        4122.0                  3.7                    0.03   
2601       10117.0                 12.1                    0.08   
2602        8663.0                 20.6                    0.04   

     female_male_ratio  year  
0                  nan  2011  
1                33:67  2011  
2                37:63  2011  
3                42:58  2011  
4                45:55  2011  
...                ...   ...  
2598             48:52  2016  
2599             36:64  2016  
2600               nan  2016  
2601             28:72  2016  
2602             43:57  2016  

[2603 rows x 14 columns]

In [109]:
# Create an empty column that will be filled in based on values in the female_male_ratio column criteria

results_fm_ratio = []

for value in df_times['female_male_ratio']:
    if value == 'nan' or value == '':
        results_fm_ratio.append(np.nan)
    elif ':' in value:
        if '100' in value:
            if '100' in value[:3]:
                results_fm_ratio.append("All Female")
            else:
                results_fm_ratio.append("All Male")
        else:
            results_fm_ratio.append(float(int(value.split(':')[0]))/float(int(value.split(':')[1])))
    else:
        results_fm_ratio.append(value[:5])
        
df_times['ratio_dec'] = results_fm_ratio

df_times

world_rank_Times                        university_name  \
0                   1                     Harvard University   
1                   2     California Institute of Technology   
2                   3  Massachusetts Institute of Technology   
3                   4                    Stanford University   
4                   5                   Princeton University   
...               ...                                    ...   
2598          601-800                    Yeungnam University   
2599          601-800            Yıldız Technical University   
2600          601-800               Yokohama City University   
2601          601-800           Yokohama National University   
2602          601-800                     Yuan Ze University   

                       country  teaching_score  international_outlook_score  \
0     United States of America            99.7                         72.4   
1     United States of America            97.7                         54.6   
2     United States of America            97.8                         82.3   
3     United States of America            98.3                         29.5   
4     United States of America            90.9                         70.3   
...                        ...             ...                          ...   
2598               South Korea            18.6                         24.3   
2599                    Turkey            14.5                         14.9   
2600                     Japan            24.0                         16.1   
2601                     Japan            20.1                         23.3   
2602                    Taiwan            16.2                         17.7   

      research_score  citations_score  income_score  overall_score_Times  \
0               98.7             98.8          34.5                 96.1   
1               98.0             99.9          83.7                 96.0   
2               91.4             99.9          87.5                 95.6   
3               98.1             99.2          64.3                 94.3   
4               95.4             99.9           NaN                 94.2   
...              ...              ...           ...                  ...   
2598            10.9             26.5          35.4                  NaN   
2599             7.6             19.3          44.0                  NaN   
2600            10.2             36.4          37.9                  NaN   
2601            16.0             13.5          40.4                  NaN   
2602            18.3             28.6          39.8                  NaN   

      num_students  student_staff_ratio  international_students  \
0          20152.0                  8.9                    0.25   
1           2243.0                  6.9                    0.27   
2          11074.0                  9.0                    0.33   
3          15596.0                  7.8                    0.22   
4           7929.0                  8.4                    0.27   
...            ...                  ...                     ...   
2598       21958.0                 15.3                    0.03   
2599       31268.0                 28.7                    0.02   
2600        4122.0                  3.7                    0.03   
2601       10117.0                 12.1                    0.08   
2602        8663.0                 20.6                    0.04   

     female_male_ratio  year ratio_dec  
0                  nan  2011       NaN  
1                33:67  2011  0.492537  
2                37:63  2011  0.587302  
3                42:58  2011  0.724138  
4                45:55  2011  0.818182  
...                ...   ...       ...  
2598             48:52  2016  0.923077  
2599             36:64  2016    0.5625  
2600               nan  2016       NaN  
2601             28:72  2016  0.388889  
2602             43:57  2016  0.754386  

[2603 rows x 15 columns]

In [110]:
# Replace the original column with the cleaned up values

df_times['female_male_ratio'] = df_times['ratio_dec']

df_times

world_rank_Times                        university_name  \
0                   1                     Harvard University   
1                   2     California Institute of Technology   
2                   3  Massachusetts Institute of Technology   
3                   4                    Stanford University   
4                   5                   Princeton University   
...               ...                                    ...   
2598          601-800                    Yeungnam University   
2599          601-800            Yıldız Technical University   
2600          601-800               Yokohama City University   
2601          601-800           Yokohama National University   
2602          601-800                     Yuan Ze University   

                       country  teaching_score  international_outlook_score  \
0     United States of America            99.7                         72.4   
1     United States of America            97.7                         54.6   
2     United States of America            97.8                         82.3   
3     United States of America            98.3                         29.5   
4     United States of America            90.9                         70.3   
...                        ...             ...                          ...   
2598               South Korea            18.6                         24.3   
2599                    Turkey            14.5                         14.9   
2600                     Japan            24.0                         16.1   
2601                     Japan            20.1                         23.3   
2602                    Taiwan            16.2                         17.7   

      research_score  citations_score  income_score  overall_score_Times  \
0               98.7             98.8          34.5                 96.1   
1               98.0             99.9          83.7                 96.0   
2               91.4             99.9          87.5                 95.6   
3               98.1             99.2          64.3                 94.3   
4               95.4             99.9           NaN                 94.2   
...              ...              ...           ...                  ...   
2598            10.9             26.5          35.4                  NaN   
2599             7.6             19.3          44.0                  NaN   
2600            10.2             36.4          37.9                  NaN   
2601            16.0             13.5          40.4                  NaN   
2602            18.3             28.6          39.8                  NaN   

      num_students  student_staff_ratio  international_students  \
0          20152.0                  8.9                    0.25   
1           2243.0                  6.9                    0.27   
2          11074.0                  9.0                    0.33   
3          15596.0                  7.8                    0.22   
4           7929.0                  8.4                    0.27   
...            ...                  ...                     ...   
2598       21958.0                 15.3                    0.03   
2599       31268.0                 28.7                    0.02   
2600        4122.0                  3.7                    0.03   
2601       10117.0                 12.1                    0.08   
2602        8663.0                 20.6                    0.04   

     female_male_ratio  year ratio_dec  
0                  NaN  2011       NaN  
1             0.492537  2011  0.492537  
2             0.587302  2011  0.587302  
3             0.724138  2011  0.724138  
4             0.818182  2011  0.818182  
...                ...   ...       ...  
2598          0.923077  2016  0.923077  
2599            0.5625  2016    0.5625  
2600               NaN  2016       NaN  
2601          0.388889  2016  0.388889  
2602          0.754386  2016  0.754386  

[2603 rows x 15 columns]

In [111]:
# Drop the newly created column

df_times = df_times.drop(columns = ['ratio_dec'])

df_times

world_rank_Times                        university_name  \
0                   1                     Harvard University   
1                   2     California Institute of Technology   
2                   3  Massachusetts Institute of Technology   
3                   4                    Stanford University   
4                   5                   Princeton University   
...               ...                                    ...   
2598          601-800                    Yeungnam University   
2599          601-800            Yıldız Technical University   
2600          601-800               Yokohama City University   
2601          601-800           Yokohama National University   
2602          601-800                     Yuan Ze University   

                       country  teaching_score  international_outlook_score  \
0     United States of America            99.7                         72.4   
1     United States of America            97.7                         54.6   
2     United States of America            97.8                         82.3   
3     United States of America            98.3                         29.5   
4     United States of America            90.9                         70.3   
...                        ...             ...                          ...   
2598               South Korea            18.6                         24.3   
2599                    Turkey            14.5                         14.9   
2600                     Japan            24.0                         16.1   
2601                     Japan            20.1                         23.3   
2602                    Taiwan            16.2                         17.7   

      research_score  citations_score  income_score  overall_score_Times  \
0               98.7             98.8          34.5                 96.1   
1               98.0             99.9          83.7                 96.0   
2               91.4             99.9          87.5                 95.6   
3               98.1             99.2          64.3                 94.3   
4               95.4             99.9           NaN                 94.2   
...              ...              ...           ...                  ...   
2598            10.9             26.5          35.4                  NaN   
2599             7.6             19.3          44.0                  NaN   
2600            10.2             36.4          37.9                  NaN   
2601            16.0             13.5          40.4                  NaN   
2602            18.3             28.6          39.8                  NaN   

      num_students  student_staff_ratio  international_students  \
0          20152.0                  8.9                    0.25   
1           2243.0                  6.9                    0.27   
2          11074.0                  9.0                    0.33   
3          15596.0                  7.8                    0.22   
4           7929.0                  8.4                    0.27   
...            ...                  ...                     ...   
2598       21958.0                 15.3                    0.03   
2599       31268.0                 28.7                    0.02   
2600        4122.0                  3.7                    0.03   
2601       10117.0                 12.1                    0.08   
2602        8663.0                 20.6                    0.04   

     female_male_ratio  year  
0                  NaN  2011  
1             0.492537  2011  
2             0.587302  2011  
3             0.724138  2011  
4             0.818182  2011  
...                ...   ...  
2598          0.923077  2016  
2599            0.5625  2016  
2600               NaN  2016  
2601          0.388889  2016  
2602          0.754386  2016  

[2603 rows x 14 columns]

In [112]:
df_times.dtypes

world_rank_Times                object
university_name                 object
country                         object
teaching_score                 float64
international_outlook_score    float64
research_score                 float64
citations_score                float64
income_score                   float64
overall_score_Times            float64
num_students                   float64
student_staff_ratio            float64
international_students         float64
female_male_ratio               object
year                             int64
dtype: object

##### Note that even after formatting this column, it will still be a mixed data type column. The reason for this is because even though most of the values are decimals, there are still some values that are strings ('All Female' or 'All Male' universities). This is fine and will have to be addressed before further analysis. 

In [113]:
# Final Check for mixed data type columns

for col in df_times.columns.tolist():
  weird = (df_times[[col]].applymap(type) != df_times[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_times[weird]) > 0:
    print (col)

female_male_ratio


As expected, the only column with mixed data types is the 'female_male_ratio' column. All other columns were addressed and formatted appropriately. 

### Addressing Missing Values

In [114]:
df_times.isnull().sum()

world_rank_Times                  0
university_name                   0
country                           0
teaching_score                    0
international_outlook_score       9
research_score                    0
citations_score                   0
income_score                    218
overall_score_Times            1402
num_students                     59
student_staff_ratio              59
international_students           67
female_male_ratio               236
year                              0
dtype: int64

There are missing values in the columns. While most of them can be ignored, it is important to note that the "overall_score_Times" column has 1402 missing values in 2603 rows. This is over 50% of the values so this column may need to be dropped later on. After looking at the universities that have these missing values, all of them had a rank that was below the top 200 so the lack of data may have been due to the rankings deeming universities outside of the top 200 unnecessary to have an overall score. 

Rows may have to be deleted that do not have an overall score, but there is still data from these universities that could be necessary so I will leave them for now. 

### Addressing Duplicate Values

In [115]:
df_dups_1 = df_times[df_times.duplicated()]

In [116]:
df_dups_1

Empty DataFrame
Columns: [world_rank_Times, university_name, country, teaching_score, international_outlook_score, research_score, citations_score, income_score, overall_score_Times, num_students, student_staff_ratio, international_students, female_male_ratio, year]
Index: []

There are no duplicate values in this dataframe.

### Summary Statistics

In [117]:
df_times.describe()

teaching_score  international_outlook_score  research_score  \
count     2603.000000                  2594.000000     2603.000000   
mean        37.801498                    52.007440       35.910257   
std         17.604218                    22.103825       21.254805   
min          9.900000                     7.100000        2.900000   
25%         24.700000                    33.425000       19.600000   
50%         33.900000                    50.300000       30.500000   
75%         46.400000                    69.000000       47.250000   
max         99.700000                   100.000000       99.400000   

       citations_score  income_score  overall_score_Times   num_students  \
count      2603.000000   2385.000000          1201.000000    2544.000000   
mean         60.921629     48.979874            59.846128   23873.758648   
std          23.073219     21.179938            12.803446   17675.946877   
min           1.200000     24.200000            41.400000     462.000000   
25%          45.500000     33.000000            50.300000   12637.750000   
50%          62.500000     41.000000            56.000000   20851.000000   
75%          79.050000     59.000000            66.200000   29991.000000   
max         100.000000    100.000000            96.100000  379231.000000   

       student_staff_ratio  international_students         year  
count          2544.000000             2536.000000  2603.000000  
mean             18.445283                0.154440  2014.075682  
std              11.458698                0.105915     1.685733  
min               0.600000                0.000000  2011.000000  
25%              11.975000                0.080000  2013.000000  
50%              16.100000                0.130000  2014.000000  
75%              21.500000                0.210000  2016.000000  
max             162.600000                0.820000  2016.000000

The counts for the columns are not all the same because of the missing values. Another thing to not is that the highest scores in the 'teaching' and 'research' categories were not equal to 100. 

### Exporting the cleaned file

In [118]:
df_times.to_csv(os.path.join(path, '02 Data','Prepared Data', 'times_cleaned_final.csv'))